FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Jobs/AzureTableMetadataStore.cs

```diff
@@ -0,0 +1,169 @@
+
+        public async Task<TriggerLeaseEntity> GetTriggerLeaseEntityAsync(byte queueType, CancellationToken cancellationToken = default)
+        {
+            return (await _metadataTableClient.GetEntityAsync<TriggerLeaseEntity>(
+                TableKeyProvider.LeasePartitionKey(queueType),
+                TableKeyProvider.LeaseRowKey(queueType),
+                cancellationToken: cancellationToken)).Value;
+        }
+
+        public async Task<Response> AddEntityAsync(ITableEntity tableEntity, CancellationToken cancellationToken = default)
+        {
+            return await _metadataTableClient.AddEntityAsync(tableEntity, cancellationToken);
+        }
+
+        public async Task<Response> UpdateEntityAsync(ITableEntity tableEntity, CancellationToken cancellationToken = default)
+        {
+            return await _metadataTableClient.UpdateEntityAsync(
+                tableEntity,
+                tableEntity.ETag,
+                cancellationToken: cancellationToken);
+        }
+
+        public async Task<CurrentTriggerEntity> GetCurrentTriggerEntityAsync(byte queueType, CancellationToken cancellationToken = default)
+        {
+            CurrentTriggerEntity entity = null;
+            try
+            {
+                var response = await _metadataTableClient.GetEntityAsync<CurrentTriggerEntity>(
+                    TableKeyProvider.TriggerPartitionKey(queueType),
+                    TableKeyProvider.TriggerRowKey(queueType),
+                    cancellationToken: cancellationToken);
+                entity = response.Value;
+            }
+            catch (RequestFailedException ex) when (ex.ErrorCode == AzureStorageErrorCode.GetEntityNotFoundErrorCode)
+            {
+                _logger.LogWarning("The current trigger doesn't exist, will create a new one.");
+            }
+            catch (Exception ex)
+            {
+                // any exceptions while getting entity will log a error and try next time
+                _logger.LogError($"Failed to get current trigger entity from table, exception: {ex.Message}");
+                throw;
+            }
+
+            return entity;
+        }
+
+        public async Task<CompartmentInfoEntity> GetCompartmentInfoEntityAsync(byte queueType, string patientId, CancellationToken cancellationToken = default)
+        {
+            return (await _metadataTableClient.GetEntityAsync<CompartmentInfoEntity>(TableKeyProvider.CompartmentPartitionKey(queueType), TableKeyProvider.CompartmentRowKey(patientId), cancellationToken: cancellationToken)).Value;
+        }
+
+        public async Task<Dictionary<string, long>> GetPatientVersionsAsync(byte queueType, List<string> patientsHash, CancellationToken cancellationToken = default)
+        {
+            var pk = TableKeyProvider.CompartmentPartitionKey(queueType);
+            var patientVersions = new Dictionary<string, long>();
+
+            for (var i = 0; i < patientsHash.Count(); i += MaxCountOfQueryEntities)
```

Comment:
Can this loop covers the whole **patientsHash**? Does the length of **patientHash** can always be divided by **MaxCountOfQueryEntities**?

Reply:
Yes, this loop will cover the whole patientsHash. If the count of the last batch is less than `MaxCountOfQueryEntities`, then it will return the remaining patients.